**Level 2: Simple Data Processing**

In [1]:
import pandas as pd

In [2]:
#Load dataset
df=pd.read_excel(r"C:\Users\Dell\Dataset1.csv.xlsx")
df.head()

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78
4,1,108,MAO,100,100,100,100,MADGOAN JN.,1,00:00:00,20:30:00,0


In [3]:
#Task 2.1 — Standardize Schedule Fields

df[['Arrival_time', 'Departure_Time']].head(20)

,Arrival_time,Departure_Time
0,00:00:00,10:25:00
1,11:06:00,11:08:00
2,11:28:00,11:30:00
3,12:10:00,00:00:00
4,00:00:00,20:30:00
5,21:04:00,21:06:00
6,21:26:00,21:28:00
7,22:25:00,00:00:00
8,19:40:00,19:40:00
9,20:18:00,20:20:00


In [4]:
df['Arrival_time'].unique()[:20]

array([datetime.time(0, 0), datetime.time(11, 6), datetime.time(11, 28),
       datetime.time(12, 10), datetime.time(21, 4), datetime.time(21, 26),
       datetime.time(22, 25), datetime.time(19, 40),
       datetime.time(20, 18), datetime.time(20, 40),
       datetime.time(21, 16), datetime.time(21, 38),
       datetime.time(21, 54), datetime.time(22, 18),
       datetime.time(22, 40), datetime.time(22, 56),
       datetime.time(23, 52), datetime.time(0, 34), datetime.time(1, 10),
       datetime.time(1, 52)], dtype=object)

In [5]:
df['Departure_Time'].unique()[:20]

array([datetime.time(10, 25), datetime.time(11, 8), datetime.time(11, 30),
       datetime.time(0, 0), datetime.time(20, 30), datetime.time(21, 6),
       datetime.time(21, 28), datetime.time(19, 40),
       datetime.time(20, 20), datetime.time(20, 42),
       datetime.time(21, 18), datetime.time(21, 40),
       datetime.time(21, 56), datetime.time(22, 20),
       datetime.time(22, 42), datetime.time(22, 58),
       datetime.time(23, 57), datetime.time(0, 36), datetime.time(1, 12),
       datetime.time(1, 54)], dtype=object)

In [6]:
#Check the current data type
df[['Arrival_time', 'Departure_Time']].dtypes

Arrival_time      object
Departure_Time    object
dtype: object

In [7]:
#Convert Arrival time
df['Arrival_time'] = pd.to_datetime(df['Arrival_time'],format='%H:%M:%S')

In [8]:
#Convert Departure Time
df['Departure_Time'] = pd.to_datetime(df['Departure_Time'],format='%H:%M:%S')

In [9]:
#Check the result
df[['Arrival_time', 'Departure_Time']].dtypes

Arrival_time      datetime64[ns]
Departure_Time    datetime64[ns]
dtype: object

In [10]:
#Look at the converted values
df[['Arrival_time', 'Departure_Time']].head(20)

,Arrival_time,Departure_Time
0,1900-01-01 00:00:00,1900-01-01 10:25:00
1,1900-01-01 11:06:00,1900-01-01 11:08:00
2,1900-01-01 11:28:00,1900-01-01 11:30:00
3,1900-01-01 12:10:00,1900-01-01 00:00:00
4,1900-01-01 00:00:00,1900-01-01 20:30:00
5,1900-01-01 21:04:00,1900-01-01 21:06:00
6,1900-01-01 21:26:00,1900-01-01 21:28:00
7,1900-01-01 22:25:00,1900-01-01 00:00:00
8,1900-01-01 19:40:00,1900-01-01 19:40:00
9,1900-01-01 20:18:00,1900-01-01 20:20:00


## Task 2.1: Standardize Schedule Fields

The `Arrival_time` and `Departure_Time` columns were initially stored as object data types.

Since these columns contain time values in `HH:MM:SS` format, they were converted into Pandas datetime format using `pd.to_datetime()`.

This standardization is necessary for performing time-based calculations, particularly for calculating total journey duration in the next task.

In [11]:
#Task 2.2 — Compute Total Journey Duration for Each Train

#Get the first and last schedule time for each train
journey_times = df.groupby('Train_No').agg(Start_Time=('Departure_Time', 'first'),
    End_Time=('Arrival_time', 'last')).reset_index()
journey_times

,Train_No,Start_Time,End_Time
0,107,1900-01-01 10:25:00,1900-01-01 12:10:00
1,108,1900-01-01 20:30:00,1900-01-01 22:25:00
2,128,1900-01-01 19:40:00,1900-01-01 17:45:00
3,290,1900-01-01 18:30:00,1900-01-01 02:30:00
4,401,1900-01-01 21:30:00,1900-01-01 10:00:00
...,...,...,...
11108,99904,1900-01-01 08:57:00,1900-01-01 09:47:00
11109,99905,1900-01-01 09:57:00,1900-01-01 10:40:00
11110,99906,1900-01-01 15:40:00,1900-01-01 16:30:00
11111,99907,1900-01-01 16:38:00,1900-01-01 17:30:00


In [12]:
#Calculate the journey duration
journey_times['Journey_Duration'] = (journey_times['End_Time'] - 
                    journey_times['Start_Time'])
journey_times.head()

,Train_No,Start_Time,End_Time,Journey_Duration
0,107,1900-01-01 10:25:00,1900-01-01 12:10:00,0 days 01:45:00
1,108,1900-01-01 20:30:00,1900-01-01 22:25:00,0 days 01:55:00
2,128,1900-01-01 19:40:00,1900-01-01 17:45:00,-1 days +22:05:00
3,290,1900-01-01 18:30:00,1900-01-01 02:30:00,-1 days +08:00:00
4,401,1900-01-01 21:30:00,1900-01-01 10:00:00,-1 days +12:30:00


In [13]:
#check data type
journey_times['Journey_Duration'].dtype

dtype('<m8[ns]')

In [14]:
#Add 24 hours to overnight journeys
journey_times.loc[journey_times['Journey_Duration'] < pd.Timedelta(0),
    'Journey_Duration'] += pd.Timedelta(days=1)
journey_times.head()

,Train_No,Start_Time,End_Time,Journey_Duration
0,107,1900-01-01 10:25:00,1900-01-01 12:10:00,0 days 01:45:00
1,108,1900-01-01 20:30:00,1900-01-01 22:25:00,0 days 01:55:00
2,128,1900-01-01 19:40:00,1900-01-01 17:45:00,0 days 22:05:00
3,290,1900-01-01 18:30:00,1900-01-01 02:30:00,0 days 08:00:00
4,401,1900-01-01 21:30:00,1900-01-01 10:00:00,0 days 12:30:00


In [15]:
#Convert duration to hours
journey_times['Journey_Duration_Hours'] = (
    journey_times['Journey_Duration'].dt.total_seconds() / 3600)
journey_times.head()

,Train_No,Start_Time,End_Time,Journey_Duration,Journey_Duration_Hours
0,107,1900-01-01 10:25:00,1900-01-01 12:10:00,0 days 01:45:00,1.750000
1,108,1900-01-01 20:30:00,1900-01-01 22:25:00,0 days 01:55:00,1.916667
2,128,1900-01-01 19:40:00,1900-01-01 17:45:00,0 days 22:05:00,22.083333
3,290,1900-01-01 18:30:00,1900-01-01 02:30:00,0 days 08:00:00,8.000000
4,401,1900-01-01 21:30:00,1900-01-01 10:00:00,0 days 12:30:00,12.500000


In [16]:
#Task 2.3 — Route Classification
df['Distance'].head()

0     0
1    32
2    49
3    78
4     0
Name: Distance, dtype: int64

In [17]:
#Check the distance values
df[['Train_No', 'Station_Name', 'Distance']].head(20)

,Train_No,Station_Name,Distance
0,107,SAWANTWADI R,0
1,107,THIVIM,32
2,107,KARMALI,49
3,107,MADGOAN JN.,78
4,108,MADGOAN JN.,0
5,108,KARMALI,33
6,108,THIVIM,51
7,108,SAWANTWADI R,83
8,128,MADGOAN JN.,0
9,128,KARMALI,33


In [18]:
df['Distance'].describe()

count    186074.000000
mean        281.353838
std         483.743964
min           0.000000
25%          23.000000
50%          73.000000
75%         291.000000
max        4260.000000
Name: Distance, dtype: float64

In [19]:
#Calculate route distance for each train
route_distance = df.groupby('Train_No')['Distance'].max().reset_index()

In [20]:
#Rename the column
route_distance = route_distance.rename(columns={'Distance': 'Route_Distance'})
route_distance.head(10)

,Train_No,Route_Distance
0,107,78
1,108,83
2,128,978
3,290,2694
4,401,1618
5,421,1276
6,422,1277
7,477,2616
8,502,1206
9,504,1313


In [21]:
#Check the distribution
route_distance['Route_Distance'].describe()

count    11113.000000
mean       348.870242
std        596.179751
min          1.000000
25%         38.000000
50%         82.000000
75%        324.000000
max       4260.000000
Name: Route_Distance, dtype: float64

In [22]:
#Create the route categories
def classify_route(distance):
    if distance <= 38:
        return 'Short'
    elif distance <= 324:
        return 'Medium'
    else:
        return 'Long'

In [23]:
#Apply the function
route_distance['Route_Category'] = route_distance['Route_Distance'].apply(
    classify_route)

route_distance.head(10)

,Train_No,Route_Distance,Route_Category
0,107,78,Medium
1,108,83,Medium
2,128,978,Long
3,290,2694,Long
4,401,1618,Long
5,421,1276,Long
6,422,1277,Long
7,477,2616,Long
8,502,1206,Long
9,504,1313,Long


In [24]:
#Count each category
route_distance['Route_Category'].value_counts()

Route_Category
Medium    5502
Short     2834
Long      2777
Name: count, dtype: int64

In [25]:
#Calculate percentages
route_distance['Route_Category'].value_counts(normalize=True) * 100

Route_Category
Medium    49.509583
Short     25.501665
Long      24.988752
Name: proportion, dtype: float64

In [26]:
#Task 2.4 — Station-wise Train Frequency

#Calculate station frequency
station_frequency = df.groupby('Station_Name').size().reset_index(
    name='Train_Frequency')
station_frequency.head(10)             

,Station_Name,Train_Frequency
0,.BAGHAJATIN,217
1,ABADA,90
2,ABHAIPUR,62
3,ABHANPUR JN.,12
4,ABHAYAPURI A,6
5,ABJUGANJ,6
6,ABOHAR,28
7,ABU ROAD,100
8,ABUTARA HALT,6
9,ACHAL GANJ,6


In [27]:
#Sort by frequency
station_frequency = station_frequency.sort_values(by='Train_Frequency',
    ascending=False)
station_frequency.head(10)           #View top 10 stations

,Station_Name,Train_Frequency
1755,CST-MUMBAI,1027
3466,KALYAN JN,828
7509,THANE,796
6834,SEALDAH,745
1592,CHENNAI BEAC,738
2950,HOWRAH JN.,699
1773,DADAR,598
2213,DUM DUM JN.,463
4245,KURLA,462
7401,TAMBARAM,434


In [28]:
#Count unique stations
df['Station_Name'].nunique()

8099

# Level 2 — Train Schedule Analysis

## Objective

The objective of Level 2 was to perform further analysis on the train schedule data by standardizing time fields, calculating journey durations, classifying train routes based on distance, and analyzing station-wise train frequency.


### Task 2.1 — Standardize Schedule Time Fields

- Converted `Arrival_time` and `Departure_Time` into appropriate time/datetime values.
- Prepared the time fields for journey duration calculations.

### Task 2.2 — Calculate Journey Duration

- Identified the first departure time and last arrival time for each train.
- Calculated the total journey duration.
- Handled overnight journeys by accounting for midnight crossings.
- Converted journey duration into hours for easier analysis.

### Task 2.3 — Route Classification

- Calculated the total route distance for each train using the maximum cumulative distance.
- Classified routes into three categories:
  - **Short:** ≤ 38 km
  - **Medium:** > 38 km and ≤ 324 km
  - **Long:** > 324 km
- Quartile-based thresholds were used because the task did not specify fixed distance boundaries.

### Task 2.4 — Station-wise Train Frequency

- Calculated the number of unique trains serving each station.
- Identified the stations with the highest train frequency.
- Found a total of **8,099 unique stations**.

## Key Findings

- The dataset contains **11,113 unique trains**.
- Journey durations were successfully calculated, including overnight journeys.
- Train routes were classified into Short, Medium, and Long categories based on route distance.
- The dataset contains **8,099 unique stations**.
- **CST-MUMBAI** had the highest train frequency, with **1,027 unique trains**.

## Conclusion

Level 2 transformed the raw train schedule data into meaningful operational insights. The analysis provided information about **journey duration, route distance, route classification, and station-wise train frequency**, creating a foundation for further analysis in the next level.